src/main/java/com/google/javascript/clutz/ImportRenameMapBuilder.java

```diff
@@ -0,0 +1,128 @@
+package com.google.javascript.clutz;
+
+import com.google.javascript.jscomp.Compiler;
+import com.google.javascript.jscomp.SourceFile;
+import com.google.javascript.rhino.Node;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Map;
+
+public class ImportRenameMapBuilder {
+  public Map<String, String> buildImportRenameMap(
+      Compiler compiler, Collection<SourceFile> sourceFiles) {
+    Map<String, String> importRenameMap = new HashMap<>();
+    for (SourceFile sf : sourceFiles) {
+      Node ast = compiler.parse(sf);
+      String moduleId = getGoogModuleId(ast);
+      // Closure can only put the symbol name into the current module's namespace, if the source
+      // file is a module, so if it isn't a module, just bail
+      if (moduleId != null) {
+        importRenameMap.putAll(buildImportRenameMap(moduleId, ast));
+      }
+    }
+    return importRenameMap;
+  }
+
+  private boolean isGoogModuleCall(Node statement) {
+    if (!statement.isExprResult()) {
+      return false;
+    }
+
+    Node expression = statement.getFirstChild();
+    return expression.isCall() && expression.getFirstChild().matchesQualifiedName("goog.module");
+  }
+
+  private boolean isGoogRequireAssignment(Node statement) {
+    if (!(statement.isConst() || statement.isVar() || statement.isLet())) {
+      return false;
+    }
+
+    if (statement.getFirstFirstChild() == null) {
+      return false;
+    }
+
+    Node rightHandSide = statement.getFirstFirstChild();
+
+    return rightHandSide.isCall()
+        && rightHandSide.getFirstChild().matchesQualifiedName("goog.require");
+  }
+
+  private boolean isGoogRequireDestructuringAssignment(Node statement) {
+    if (!(statement.isConst() || statement.isVar() || statement.isLet())) {
+      return false;
+    }
+
+    if (!statement.getFirstChild().isDestructuringLhs()) {
+      return false;
+    }
+
+    Node destructuringAssignment = statement.getFirstChild();
+
+    Node rightHandSide = destructuringAssignment.getChildAtIndex(1);
+
+    return rightHandSide.isCall()
+        && rightHandSide.getFirstChild().matchesQualifiedName("goog.require");
+  }
+
+  private String getGoogModuleId(Node astRoot) {
+    if (astRoot.getFirstChild() == null || !astRoot.getFirstChild().isModuleBody()) {
+      return null;
+    }
+
+    Node moduleBody = astRoot.getFirstChild();
+    for (Node statement : moduleBody.children()) {
+      if (isGoogModuleCall(statement)) {
+        return statement.getFirstChild().getChildAtIndex(1).getString();
+      }
+    }
+    return null;
+  }
+
+  /**
+   * If we're running in incremental mode, closure gives imported symbols names in the current
+   * module's namespace, instead of the namespace of the module we're importing from. This reads the
```

Comment:
I think this text would be easier to understand with an example of the names.

Reply:
Done.